src/main/scala/mesosphere/util/CapConcurrentExecutions.scala

```diff
@@ -0,0 +1,157 @@
+package mesosphere.util
+
+import javax.annotation.PreDestroy
+
+import akka.actor._
+import akka.pattern.pipe
+import mesosphere.marathon.metrics.Metrics.AtomicIntGauge
+import mesosphere.marathon.metrics.{ MetricPrefixes, Metrics }
+import mesosphere.util.RestrictParallelExecutionsActor.Finished
+import org.slf4j.LoggerFactory
+
+import scala.collection.immutable.Queue
+import scala.concurrent.Future
+import scala.util.Failure
+import scala.util.control.NonFatal
+
+/**
+  * Allows capping parallel executions of methods which return `scala.concurrent.Future`s.
+  * Only `maxParallel` concurrent executions are allowed.
+  *
+  * {{{
+  * scala> import mesosphere.util.CapConcurrentExecutions
+  * scala> val capParallelExecution = CapConcurrentExecutions(someActorRefFactory, "serialize")
+  * scala> def myFutureReturningFunc: Future[Int] = Future.successful(1)
+  * scala> val result: Future[Int] = capParallelExecution(myFutureReturningFunc)
+  * }}}
+  */
+object CapConcurrentExecutions {
+  private val log = LoggerFactory.getLogger(getClass.getName)
+
+  def apply[T](
+    metrics: CapConcurrentExecutionsMetrics,
+    actorRefFactory: ActorRefFactory,
+    actorName: String,
+    maxParallel: Int,
+    maxQueued: Int): CapConcurrentExecutions = {
+    new CapConcurrentExecutions(metrics, actorRefFactory, actorName, maxParallel, maxQueued)
+  }
+}
+
+class CapConcurrentExecutionsMetrics(metrics: Metrics, metricsClass: Class[_]) {
+  val queued = metrics.gauge(metrics.name(MetricPrefixes.SERVICE, metricsClass, "queued"), new AtomicIntGauge)
+  val processing = metrics.gauge(metrics.name(MetricPrefixes.SERVICE, metricsClass, "processing"), new AtomicIntGauge)
+  val processingTimer = metrics.timer(metrics.name(MetricPrefixes.SERVICE, metricsClass, "processing-time"))
+
+  def reset(): Unit = {
+    queued.setValue(0)
+    processing.setValue(0)
+  }
+}
+
+class CapConcurrentExecutions private (
+    metrics: CapConcurrentExecutionsMetrics,
+    actorRefFactory: ActorRefFactory,
+    actorName: String,
+    maxParallel: Int,
+    maxQueued: Int) {
+  import CapConcurrentExecutions.log
+
+  private[this] val serializeExecutionActorProps =
+    RestrictParallelExecutionsActor.props(metrics, maxParallel = maxParallel, maxQueued = maxQueued)
+  private[this] val serializeExecutionActorRef = actorRefFactory.actorOf(serializeExecutionActorProps, actorName)
+
+  def apply[T](block: => Future[T]): Future[T] = {
```

Comment:
Shouldn't this also have an EC as implicit parameter?
On which EC is that block executed?